### ISSUE TITLE
Is there anyway to help me speed up in loading the video data?

### ISSUE THREAD

**Issue by marco-hmc** [facebookresearch/pytorchvideo#45]
I'm writing a dataloaer file with pytorchvideo currently, all the thing is working quite fine except only the speed. It takes me almost 1 second to load a video at one time. Some of my codes are shown below.

    transform = ApplyTransformToKey(
        key="video",
        transform=Compose([
            UniformTemporalSubsample(64),
            Lambda(lambda x: x / 255.0),
            NormalizeVideo(mean, std),
            ShortSideScale(size=256),
            CenterCropVideo(256),
        ]),
    )
> My transform, in simple, getting 64 frames in 64 second-length video clip.

    def __getitem__(self, index):
        video_data = self.video.get_clip(start_sec=index * self.length, end_sec=(index + 1) * self.length)
        video_data = self.transform(video_data)
        imgs_data, audios_data = video_data['video'], video_data['audio']
        labels = torch.from_numpy(self.va_read(int(index + 1))).t()
        return imgs_data, audios_data, labels, self.movie_name, index

>part of the consumming time of getting the data. The n
Epoch: [0][4/24]  Time_train(average) 30.307 (30.044)  Data_load(average) 30.035 (29.764) 
Epoch: [0][5/24]  Time_train(average) 58.678 (31.728)  Data_load(average) 58.383 (31.448) 
Epoch: [0][6/24]  Time_train(average) 49.739 (35.320)  Data_load(average) 49.315 (35.036) 
Epoch: [0][6/24]  Time_train(average) 53.579 (38.151)  Data_load(average) 53.336 (37.872) 
Epoch: [0][7/24]  Time_train(average) 51.027 (38.822)  Data_load(average) 50.760 (38.543)

So is there anything I can do for speeding up the loading video? One more thing is that given the fixth length of video clip, the length of audios seems is not the same which will cause an error if the batch size is more than 1, how can i set the sampling rate of audios? I will be thankful for all kinds of help or suggestions.

**Comment by kalyanvasudev** [comment]
@tullie , could you please take a loot at this?

**Comment by kalyanvasudev** [comment]
> @tullie , could you please take a loot at this?

I wonder if this is the same behavior we see in this issue - https://github.com/facebookresearch/pytorchvideo/issues/24

**Comment by marco-hmc** [comment]
> > @tullie , could you please take a loot at this?
> 
> I wonder if this is the same behavior we see in this issue - #24

I have two mainly questions indeed.

**Comment by tullie** [comment]
Hi @marco-hmc,

For video training, data loading is typically the performance bottleneck so seeing the majority of time loading video data isn't surprising. Saying that though, there's ways to improve performance which you may not have considered yet:

1. If decoding speed is more expensive than reading speed, you can decode before training using ffmpeg. If you store the videos as a folder of png/jpegs, then PyTorchVideo LabeledVideoDataset will read it without decoding. 
2. I'd suggest preprocessing all the videos as much as possible. E.g. if you're only training with 256x256 resolution videos but you have 800x1000 videos, then it's worth doing that dowsample before training. Likewise for frame rate resampling.
3. If you do stick with decoding during training, it's important that you're using selective decoding (i.e. you're only decoding the part of the video you want to train with at each step). The decoder can only use this, if the duration and timebase information is in the video header. If it isn't you'll have to add this in a preprocessing step.
4. Make sure you're using multiple workers in your PyTorch DataLoader, typically set num_workers=<num_cpus>, is good. You can also adjust the process prefetch number in the DataLoader.

Regarding your second problem, you need to add a transform to resample the modality types to a constant value. You can see an example of how this is done for both audio and video in this classification example https://github.com/facebookresearch/pytorchvideo/blob/master/tutorials/video_classification_example/train.py

Thanks,
Tullie

**Comment by marco-hmc** [comment]
> Hi @marco-hmc,
> 
> For video training, data loading is typically the performance bottleneck so seeing the majority of time loading video data isn't surprising. Saying that though, there's ways to improve performance which you may not have considered yet:
> 
> 1. If decoding speed is more expensive than reading speed, you can decode before training using ffmpeg. If you store the videos as a folder of png/jpegs, then PyTorchVideo LabeledVideoDataset will read it without decoding.
> 2. I'd suggest preprocessing all the videos as much as possible. E.g. if you're only training with 256x256 resolution videos but you have 800x1000 videos, then it's worth doing that dowsample before training. Likewise for frame rate resampling.
> 3. If you do stick with decoding during training, it's important that you're using selective decoding (i.e. you're only decoding the part of the video you want to train with at each step). The decoder can only use this, if the duration and timebase information is in the video header. If it isn't you'll have to add this in a preprocessing step.
> 4. Make sure you're using multiple workers in your PyTorch DataLoader, typically set num_workers=<num_cpus>, is good. You can also adjust the process prefetch number in the DataLoader.
> 
> Regarding your second problem, you need to add a transform to resample the modality types to a constant value. You can see an example of how this is done for both audio and video in this classification example https://github.com/facebookresearch/pytorchvideo/blob/master/tutorials/video_classification_example/train.py
> 
> Thanks,
> Tullie

Thanks for your kindly help. All of your suggestion gonna help a lot.
